### PR TITLE
[ProductVariant][Product][Admin] Unique code during generation of product's variants and during adding product's images

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | no |
| Related tickets | - |
| License         | MIT |

### CODE REVIEW

File: features/product/managing_product_variants/generating_product_variant_validation.feature [hunk 1]
```diff
@@ -25,3 +25,36 @@ Feature: Generating product variant generation
         And I try to generate it
         Then I should be notified that code is required for the 1st variant
         And I should see 0 variants in the list
+
+    @ui
+    Scenario: Generation a product variants without specific required fields for second variant
```
Comment (NeverResponse) [inline features/product/managing_product_variants/generating_product_variant_validation.feature]: `Generating product variants` ... and so on

File: features/product/managing_product_variants/generating_product_variant_validation.feature [hunk 2]
```diff
@@ -25,3 +25,36 @@ Feature: Generating product variant generation
...
+        Given I want to generate new variants for this product
```
Comment (GSadee) [inline features/product/managing_product_variants/generating_product_variant_validation.feature]: `When...`

File: src/Sylius/Behat/Page/Admin/Product/UpdateSimpleProductPage.php [hunk 3]
```diff
@@ -360,6 +360,23 @@ public function getPriceForChannel($channelName)
     /**
      * {@inheritdoc}
      */
+    public function getValidationMessageForImageAtPlace($place)
```
Comment (NeverResponse) [inline src/Sylius/Behat/Page/Admin/Product/UpdateSimpleProductPage.php]: Wouldn't position be better? Place looks... well... weird.

File: features/product/managing_product_variants/generating_product_variant_validation.feature [hunk 4]
```diff
@@ -25,3 +25,36 @@ Feature: Generating product variant generation
...
+        When I specify that the 1st variant is identified by "WYBOROWA_ORANGE" code
+        And I specify that the 1st variant costs "$90" in "United States" channel
+        And I try to generate it
+        Then I should be notified that code is required for the 2st variant
+        Then I should be notified that prices in all channels must be defined for the 2st variant
+        And I should see 0 variants in the list
```
Comment (GSadee) [inline features/product/managing_product_variants/generating_product_variant_validation.feature]: `And I should not see variants in the list`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 5]
```diff
@@ -425,12 +425,23 @@ public function iShouldBeNotifiedThatCodeIsRequiredForVariant($position)
...
         );
     }
 
     /**
+     * @Then /^I should be notified that variant code must be unique within this product for the (\d)(?:st|nd|rd|th) variant$/
+     */
+    public function iShouldBeNotifiedThatVariantCodeMustBeUniqueWithinThisProductForYheVariant($position)
+    {
+        Assert::same(
+            $this->generatePage->getValidationMessage('code', $position -1),
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `$position - 1`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 6]
```diff
@@ -648,15 +659,15 @@ public function iSpecifyThereAreVariantsIdentifiedByCodeWithCost($nthVariant, $c
      */
     public function iSpecifyThereAreVariantsIdentifiedByCode($nthVariant, $code)
     {
-        $this->generatePage->nameCode($nthVariant, $code - 1);
+        $this->generatePage->nameCode($nthVariant -1, $code);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: CS `$nthVariant - 1`

File: src/Sylius/Bundle/ResourceBundle/spec/Validator/WithinCollectionUniqueCodeValidatorSpec.php [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        $context->buildViolation(Argument::type('string'))->shouldBeCalled()->willReturn($violationBuilder);
+        $violationBuilder->atPath('[0].code')->shouldBeCalled()->willReturn($violationBuilder);
+
+        $context->buildViolation(Argument::type('string'))->shouldBeCalled()->willReturn($violationBuilder);
+        $violationBuilder->atPath('[2].code')->shouldBeCalled()->willReturn($violationBuilder);
+
+        $violationBuilder->addViolation()->shouldBeCalledTimes(2);
+
+
+        $this->validate([ $firstEntity->getWrappedObject(), $secondEntity->getWrappedObject(),  $thirdEntity->getWrappedObject()], $constraint);
```
Comment (GSadee) [inline src/Sylius/Bundle/ResourceBundle/spec/Validator/WithinCollectionUniqueCodeValidatorSpec.php]: Too long line

File: UPGRADE.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Table of Contents
+
+- [CoreBundle](#corebundle)
+
+### CoreBundle
+
+* ImageUniqueCode and ImageUniqueCodeValidator was deleted and replaced by WithinCollectionUniqueCodeValidator, WithinCollectionUniqueCode form ResourceBundle.
+  To use it replace name of constraint in constraint mapping file from:
+  &nbsp;&nbsp;`Sylius\Bundle\CoreBundle\Validator\Constraints\ImageUniqueCode`
```
Comment (pjedrzejewski) [inline UPGRADE.md]: Please remove these nbsps.

File: UPGRADE.md [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Table of Contents
```
Comment (pjedrzejewski) [inline UPGRADE.md]:
```
UPGRADE FROM 1.0.0-beta.1 to 1.0.0-beta.2
================================

# Table of Contents
```

File: src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($collectionOfEntities, Constraint $constraint)
+    {
+        $collectionOfEntitiesCodes = [];
+
+        /** @var CodeAwareInterface $entity */
+        foreach ($collectionOfEntities as $key => $entity) {
+            if(null === $entity->getCode()) {
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php]: CS!

File: src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+namespace Sylius\Bundle\ResourceBundle\Validator;
+
+use Sylius\Component\Resource\Model\CodeAwareInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class WithinCollectionUniqueCodeValidator extends ConstraintValidator
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php]: Perhaps we can name it ``UniqueWithinCollectionConstraint`` and use property accessor to make the field that is supposed to be unique configurable? This way we could use that for zones for example and not only for codes.

File: features/product/managing_product_variants/generating_product_variant_validation.feature [hunk 12]
```diff
@@ -11,17 +11,50 @@ Feature: Generating product variant generation
...
-        And I should see 0 variants in the list
+        And I should not see variants in the list
+
+    @ui
+    Scenario: Generating product's variants without specific required fields for second variant
+        When I want to generate new variants for this product
+        And I specify that the 1st variant is identified by "WYBOROWA_ORANGE" code
+        And I specify that the 1st variant costs "$90" in "United States" channel
+        And I try to generate it
+        Then I should be notified that code is required for the 2st variant
```
Comment (michalmarcinkowski) [inline features/product/managing_product_variants/generating_product_variant_validation.feature]: `2st` -> `2nd`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 13]
```diff
@@ -414,7 +415,7 @@ public function iShouldBeNotifiedThatThisVariantAlreadyExists()
     public function iShouldBeNotifiedThatCodeIsRequiredForVariant($position)
     {
         Assert::same(
-            $this->generatePage->getValidationMessage('code', $position),
+            $this->generatePage->getValidationMessage('code', $position -1),
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: Missing space before number `$position - 1`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 14]
```diff
@@ -648,15 +660,15 @@ public function iSpecifyThereAreVariantsIdentifiedByCodeWithCost($nthVariant, $c
      */
     public function iSpecifyThereAreVariantsIdentifiedByCode($nthVariant, $code)
     {
-        $this->generatePage->nameCode($nthVariant, $code - 1);
+        $this->generatePage->nameCode($nthVariant - 1, $code);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `nameCode`? Should be `specifyCode`?

File: UPGRADE.md (hunk 9 above)
Comment (tuka217) [inline UPGRADE.md]: Done

File: src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php (hunk 11 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php]: @tuka217 Please rename it (and update UPGRADE file) and it is good to go.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 15]
```diff
@@ -648,15 +660,15 @@ public function iSpecifyThereAreVariantsIdentifiedByCodeWithCost($nthVariant, $c
...
+        $this->generatePage->specifyCode($nthVariant - 1, $code);
     }
 
     /**
      * @When /^I specify that the (\d)(?:st|nd|rd|th) variant costs "(?:€|£|\$)([^"]+)" in ("[^"]+") channel$/
      */
     public function iSpecifyThereAreVariantsWithCost($nthVariant, $price, $channelName)
     {
-        $this->generatePage->specifyPrice($nthVariant, $price - 1, $channelName);
+        $this->generatePage->specifyPrice($nthVariant -1, $price, $channelName);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: CS

File: UPGRADE.md [hunk 16]
```diff
@@ -0,0 +1,12 @@
+UPGRADE FROM 1.0.0-beta.1 to 1.0.0-beta.2
+================================
+
+# Table of Contents
+
+- [CoreBundle](#corebundle)
+
+### CoreBundle
+
+* ImageUniqueCode and ImageUniqueCodeValidator was deleted and replaced by WithinCollectionUniqueCodeValidator, WithinCollectionUniqueCode form ResourceBundle.
```
Comment (pamil) [inline UPGRADE.md]: Let's write the class names using monospace font. Also "was" -> "were", "form" -> "from" (and the order of classes is different 🌮 ).

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php [hunk 17]
```diff
@@ -999,6 +999,19 @@ public function thisProductShouldNoLongerHavePriceForChannel(ProductInterface $p
     }
 
     /**
+     * @Then I should be notified that the :imageNumber image should have an unique code
+     */
+    public function iShouldBeNotifiedThatTheFirstImageShouldHaveAnUniqueCode($imageNumber)
+    {
+        preg_match_all('!\d+!', $imageNumber, $matches);
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: What is

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php [hunk 18]
```diff
@@ -999,6 +999,19 @@ public function thisProductShouldNoLongerHavePriceForChannel(ProductInterface $p
...
+
+        Assert::same(
+            $this->updateSimpleProductPage->getValidationMessageForImageAtPosition(((int) $matches[0][0]) - 1),
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: going on here?

Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: 🌮

Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductsContext.php]: `((int) $matches[0][0]) - 1` -> `((int) $imageNumber) - 1`

File: src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php (hunk 11 above)
Comment (tuka217) [inline src/Sylius/Bundle/ResourceBundle/Validator/WithinCollectionUniqueCodeValidator.php]: It is update, open this file.

**Comment by pjedrzejewski** [comment]
Thank you Ania!

**Pull request merged**